x/merkledb/proof.go

```diff
@@ -138,31 +148,39 @@ func (proof *Proof) Verify(ctx context.Context, expectedRootID ids.ID, tokenSize
 	if len(proof.Path) == 0 {
 		return ErrEmptyProof
 	}
-
+	if proof.Key.hasPartialByte() {
+		return ErrProofKeyPartialByte
+	}
 	if err := verifyProofPath(proof.Path, maybe.Some(proof.Key)); err != nil {
 		return err
 	}
 
-	// Confirm that the last proof node's value matches the claimed proof value
 	lastNode := proof.Path[len(proof.Path)-1]
+	isInclusionProof := lastNode.Key == proof.Key
 
-	// If the last proof node's key is [proof.Key] (i.e. this is an inclusion proof)
-	// then the value of the last proof node must match [proof.Value].
-	// Note partial byte length keys can never match the [proof.Key] since it's bytes,
-	// and thus has a whole number of bytes
-	if !lastNode.Key.hasPartialByte() &&
-		proof.Key == lastNode.Key &&
-		!valueOrHashMatches(proof.Value, lastNode.ValueOrHash) {
-		return ErrProofValueDoesntMatch
-	}
-
-	// If the last proof node has a length not evenly divisible into bytes or a different key than [proof.Key]
-	// then this is an exclusion proof and should prove that [proof.Key] isn't in the trie.
-	// Note length not evenly divisible into bytes can never match the [proof.Key] since it's bytes,
-	// and thus an exact number of bytes.
-	if (lastNode.Key.hasPartialByte() || proof.Key != lastNode.Key) &&
-		proof.Value.HasValue() {
-		return ErrProofValueDoesntMatch
+	switch {
+	case isInclusionProof && !valueOrHashMatches(proof.Value, lastNode.ValueOrHash):
+		return ErrInclusionProofWrongValue
+	case !isInclusionProof && proof.Value.HasValue():
+		return ErrExclusionProofUnexpectedValue
+	case !isInclusionProof && proof.Key.HasPrefix(lastNode.Key):
+		// The last node in an exclusion proof must be either:
+		// 1. The node that would be the parent of [proof.Key] if [proof.Key] were in the trie, or
+		// 2. The node that is where [proof.Key] would be if [proof.Key] were in the trie
+		// We are in case 1. Make sure there's no child which is a prefix of [proof.Key].
+		unmatchedKey := proof.Key.Skip(lastNode.Key.length)
+		if unmatchedKey.Length() < tokenSize {
+			// The last node's key is a prefix of [proof.Key], but isn't equal to it,
+			// so there should be at least one unmatched token.
+			return ErrExpectedUnmatchedToken
+		}
+		nextUnmatchedToken := unmatchedKey.Token(0, tokenSize)
+		if _, ok := lastNode.Children[nextUnmatchedToken]; ok {
+			// The proof path should include all nodes which are prefixes of [proof.Key]
+			// but there's another child of the last node which is a prefix of [proof.Key].
+			// That child (or its descendant) should be the last node in the proof path.
+			return ErrUnexpectedEndOfProof
```

Comment:
This fixes a bug in the old code. Previously, it was possible for an exclusion proof to pass verification even if the key really _is_ in the trie. We were not checking that the proof path extends as far down as possible.

To generate such a proof in the old code:
1. Generate a valid inclusion proof
2. Remove the last node from `proof.Path` 
3. Set `Value` to `Nothing`

We would insert the parent of the node at the key, as well as its children IDs, and as a result verification would pass because the root ID was correct. In other words, when we're verifying an exclusion proof, we need to assert that it's impossible for the key we're proving to exist "below" the nodes in the proof.